### PR TITLE
fix: height calculation of BaseTextInput

### PR DESCRIPTION
**Pull request by dummy-1111**
### Details

Correct the height calculation of BaseTextInput

### Fixed Issues

$ https://github.com/Expensify/App/issues/21271
PROPOSAL: https://github.com/Expensify/App/issues/21271#issuecomment-1628784442

### Tests

1. Open ND and log in with any account
2. Click FAB and Assign Task
3. Enter first line in description field
4. Press Enter key
5. Verify that the cursor is fully visible

- [x] Verify that no errors appear in the JS console

### Offline tests

1. Open ND and log in with any account
2. Click FAB and Assign Task
3. Enter first line in description field
4. Press Enter key
5. Verify that the cursor is fully visible

### QA Steps
1. Open ND and log in with any account
2. Click FAB and Assign Task
3. Enter first line in description field
4. Press Enter key
5. Verify that the cursor is fully visible

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/126839717/776fccc6-f7fe-492d-bc7a-e2a6d3997d8f

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/126839717/3bea81aa-3078-4f07-8723-4b0db3064944

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/126839717/d25ee9ba-bddb-48bf-a23e-7c4295869638

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/126839717/7e555e25-82a1-4fa6-8b9d-a2632dfc6f76

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/126839717/92b8a441-3e39-4925-b559-06bdc671d30e

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/126839717/d3ae4ba3-0442-4980-9846-24f01b0a6378

</details>

### CODE REVIEW

**Comment by s-alves10** [comment]
I've chosen `\u200B`(Zero-width space character) based on tests on various platforms.

**Comment by s-alves10** [comment]
@sobitneupane 

Please take a look. thanks

File: src/components/TextInput/BaseTextInput.js [hunk 1]
```diff
@@ -395,7 +395,7 @@ function BaseTextInput(props) {
                         setTextInputHeight(e.nativeEvent.layout.height);
                     }}
                 >
-                    {props.value || props.placeholder}
+                    {props.value ? `${props.value}\u200B` : props.placeholder}
```
Comment (sobitneupane) [inline src/components/TextInput/BaseTextInput.js:398]: Let's add some comment to explain why the change was needed.

@s-alves10 Why did we use `\u200B` instead of `\n`?

[the remark this answers is not in the document]
Comment (s-alves10) [reply src/components/TextInput/BaseTextInput.js:398]: https://github.com/Expensify/App/pull/23286#issuecomment-1644174450

[the remark this answers is not in the document]
Comment (sobitneupane) [reply src/components/TextInput/BaseTextInput.js:398]: @s-alves10 I saw the comment. 

> I've chosen \u200B(Zero-width space character) based on tests on various platforms.

Can you please explain. In which platform do we have the issue? And what issue do we have?

[the remark this answers is not in the document]
Comment (s-alves10) [reply src/components/TextInput/BaseTextInput.js:398]: Yes, I saw that adding \n increases the height in ios native. So I've done some research work about zero width character and found ZWSC(\u200B)

[the remark this answers is not in the document]
Comment (s-alves10) [reply src/components/TextInput/BaseTextInput.js:398]:
```
Need to add \u200B(Zero-width space character) to get correct layout in cases value is ended with \n(EOL)
```
Is this comment ok?

[the remark this answers is not in the document]
Comment (sobitneupane) [reply src/components/TextInput/BaseTextInput.js:398]: There is an issue on Android. An extra line is added if we enter any character.

https://github.com/Expensify/App/assets/25876548/177c820b-43bd-4a4f-9811-65d5b4142e46

[the remark this answers is not in the document]
Comment (sobitneupane) [reply src/components/TextInput/BaseTextInput.js:398]: The other thing that I didn't understand is: Shouldn't we add the character only when user presses (i.e., line ends with `\n`)?

[the remark this answers is not in the document]
Comment (s-alves10) [reply src/components/TextInput/BaseTextInput.js:398]: https://github.com/Expensify/App/issues/21271#issuecomment-1643183002
I used alternative solution
Let me check again

[the remark this answers is not in the document]
Comment (s-alves10) [reply src/components/TextInput/BaseTextInput.js:398]: @sobitneupane 
I tested again but it work on all platforms. Can you check again?

The current change is based on the fact that ZWSC doesn't occupy space but it's still a character. \n character at the last position doesn't occupy space on some platforms, but if that is in the middle of the string, it will increase height on all platforms

[the remark this answers is not in the document]
Comment (s-alves10) [reply src/components/TextInput/BaseTextInput.js:399]: comment added

[the remark this answers is not in the document]
Comment (sobitneupane) [reply src/components/TextInput/BaseTextInput.js:399]: > ```
> Need to add \u200B(Zero-width space character) to get correct layout in cases value is ended with \n(EOL)
> ```
> 
> Is this comment ok?

```\u200B added to solve the issue(https://github.com/Expensify/App/issues/21271): "Cursor not visible on hitting enter"```

[the remark this answers is not in the document]
Comment (sobitneupane) [reply src/components/TextInput/BaseTextInput.js:399]: > The other thing that I didn't understand is: Shouldn't we add the character only when user presses (i.e., line ends with \n)?

I cannot find answer to this question in proposal as well. Can you please help me understand it?

[the remark this answers is not in the document]
Comment (s-alves10) [reply src/components/TextInput/BaseTextInput.js:399]: comment updated

[the remark this answers is not in the document]
Comment (sobitneupane) [reply src/components/TextInput/BaseTextInput.js:399]: > I tested again but it works fine on all platforms. Can you check again?

I am bit surprised that I cannot reproduce it any longer.

>This adds the space character when the string ends with '\n' to prevent the string to be ended with '\n'. But we can solve this issue by simply adding ZWSC because this has a zero-width and assures that the string won't be ended with '\n'. At first, I suggested to use '\n', but it has a problem in native platforms, so I decided to use ZWSC

So, there is no solid reason to do so. Why don't we add zwsp character only if the value ends with `\n`.

[the remark this answers is not in the document]
Comment (sobitneupane) [reply src/components/TextInput/BaseTextInput.js:399]: @s-alves10 Please let me know what you think about above comment.

[the remark this answers is not in the document]
Comment (s-alves10) [reply src/components/TextInput/BaseTextInput.js:399]: > Why don't we add zwsp character only if the value ends with `\n`

The issue happens when the value ends with `\n`.
You're right. We can solve this by adding extra ZWSC when the value ends with `\n`. But we can add ZWSC when the value doesn't end with `\n` as well because ZWSC has zero-width. The reason is for simplicity.

[the remark this answers is not in the document]
Comment (s-alves10) [reply src/components/TextInput/BaseTextInput.js:399]: @sobitneupane 

Should I update the code?

[the remark this answers is not in the document]
Comment (sobitneupane) [reply src/components/TextInput/BaseTextInput.js:399]: > Should I update the code?

Yes, let's add the character only for necessary case.

[the remark this answers is not in the document]
Comment (s-alves10) [reply src/components/TextInput/BaseTextInput.js]: @sobitneupane 

Updated. Will you take a look again?

[the remark this answers is not in the document]
Comment (sobitneupane) [reply src/components/TextInput/BaseTextInput.js]: @s-alves10 Can you please update the screenshots as well.

[the remark this answers is not in the document]
Comment (s-alves10) [reply src/components/TextInput/BaseTextInput.js]: Screenshots are the same as before. What's it for?

[the remark this answers is not in the document]
Comment (sobitneupane) [reply src/components/TextInput/BaseTextInput.js]: > Screenshots are the same as before. I've verified that it works fine. Do I need to record them again?

@s-alves10 Yes.

[the remark this answers is not in the document]
Comment (s-alves10) [reply src/components/TextInput/BaseTextInput.js]: @sobitneupane 
Videos updated

**Review by sobitneupane — APPROVED** [review]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/25876548/e5716fe5-5fc4-4290-9c55-3faef0639fc5

</details>

<details>
<summary>Mobile Web - Chrome</summary>



</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/25876548/67a44244-2c8e-4fdf-83b3-625636d57da4

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/25876548/7c5911d6-2482-43ef-a7a0-de494a43bc85

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/25876548/e1995241-dcb6-4f70-b518-5a03849d0303

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/25876548/d3c3dbbf-36ba-4562-adc6-50b576e1ecab

</details>

File: src/components/TextInput/BaseTextInput.js [hunk 2]
```diff
@@ -395,7 +395,8 @@ function BaseTextInput(props) {
                         setTextInputHeight(e.nativeEvent.layout.height);
                     }}
                 >
-                    {props.value || props.placeholder}
+                    {/* \u200B added when value ends with '\n' to solve the issue(https://github.com/Expensify/App/issues/21271): "Cursor not visible on hitting enter" */}
```
Comment (dangrous) [inline src/components/TextInput/BaseTextInput.js:398]:
```suggestion
                    {/* \u200B added to solve the issue of not expanding the text input enough when the value ends with '\n' (https://github.com/Expensify/App/issues/21271) */}
```

**Review by dangrous — CHANGES_REQUESTED** [review]
Code looks good, I made a suggestion to clarify the comment a little more

File: src/components/TextInput/BaseTextInput.js (hunk 2 above)
Comment (s-alves10) [reply src/components/TextInput/BaseTextInput.js]: @dangrous 
Comment updated

**Review by dangrous — APPROVED** [review]
